product_barcode_constraint_per_company/models/product_product.py

```diff
@@ -2,24 +2,22 @@
 # @author: Sylvain LE GAL (https://twitter.com/legalsylvain)
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
 
-from odoo import _, fields, models
+from odoo import fields, models
 
 
 class ProductProduct(models.Model):
     _inherit = "product.product"
 
-    tmpl_company_id = fields.Many2one(
-        related="product_tmpl_id.company_id",
-        string="Company of the related Template",
-        store=True,
+    # Store related field
+    company_id = fields.Many2one(
```

Comment:
I would say, this could be in another module to avoid collisions

Reply:
Maybe I don't see where is the risk of collision.

The add-on is only 10 lines of Python : it stores the related field which already exist in standard, and monkey patch the SQL constraint. Nothing more.

Reply:
@rousseldenis in which repo should that module be published in your opinion?

Reply:
> @florentx I wanted to say the functionality of storing company_id on product.product level could be seen as independent feature (but base one for this).

@rousseldenis @simahawk would https://github.com/OCA/product-variant work in your opinion?

Reply:
> @florentx I wanted to say the functionality of storing company_id on product.product level could be seen as independent feature (but base one for this).

Sorry, but I'm not sure I got your point here, if you mean that the code related to the functionality of storing company_id at product.product level should be put in a separate module, then @florentx  module would then only need to handle  _sql_constraints.
So for only ten lines of code we will create two modules, it seems a bit complicated.

@rousseldenis @simahawk Could you please elaborate on this further?

Reply:
@simahawk ping :)

Reply:
I'd go for product-attribute

Reply:
@simahawk ok we'll push a simple module with just those lines

Reply:
@florentx please let us know if you have time to finish this PR, we're available to pick it up and bring it to completion if that's ok for you :)